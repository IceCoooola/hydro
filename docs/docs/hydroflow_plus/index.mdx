---
sidebar_position: 0
---

# Introduction
Hydroflow+ layers a high-level Rust API over the Hydroflow IR, making it possible to write dataflow programs that span multiple processes with straightline, functional Rust code. Hydroflow+ is built on top of [Stageleft](./stageleft.mdx), which allows Hydroflow+ to emit regular Hydroflow programs that are compiled into efficient Rust binaries. It also integrates with [Hydro Deploy](../deploy/index.md) to make it easy to deploy and run Hydroflow+ programs on a cluster.

The main logic of Hydroflow+ programs manipulates **streams**, which capture infinite ordered sequences of elements. Streams are transformed using classic functional operators such as `map`, `filter`, and `fold`, as well as relational operators such as `join`. To build **distributed** dataflow programs, Hydroflow+ also introduces the concept of **processes**, which capture _where_ a stream is being processed.

## Setup
Hydroflow+ requires a particular workspace setup, as any crate that uses Hydroflow+ must have an supporting macro crate to drive the code generation. To get started, we recommend using the Hydroflow+ template.

```bash
#shell-command-next-line
cargo install cargo-generate
#shell-command-next-line
cargo generate hydro-project/hydroflow-plus-template
```

Then, you can test the example dataflow:

```bash
#shell-command-next-line
cargo test
```
