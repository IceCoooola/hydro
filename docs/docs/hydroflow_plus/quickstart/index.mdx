# Quickstart
In this tutorial, we'll walk through the basics of Hydroflow+ by building a simple dataflow that prints out the first 10 natural numbers. We'll start with a single process, then pipeline the computation, and finally distribute it across a cluster.

To get started with a new project, we'll use the Hydroflow+ template. The template comes with a pre-configured build system and an implementation of the following examples.

```bash
#shell-command-next-line
cargo install cargo-generate
#shell-command-next-line
cargo generate hydro-project/hydroflow-plus-template
# Project Name: my-example-project
#shell-command-next-line
cd my-example-project
```
