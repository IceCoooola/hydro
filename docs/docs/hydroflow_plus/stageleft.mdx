---
title: Stageleft
sidebar_position: 6
---

import StageleftDocs from '../../../stageleft/README.md'

<StageleftDocs/>
