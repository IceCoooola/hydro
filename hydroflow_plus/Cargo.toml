[package]
name = "hydroflow_plus"
publish = true
version = "0.7.0"
edition = "2021"
license = "Apache-2.0"
documentation = "https://docs.rs/hydroflow_plus/"
description = "Functional programming API for hydroflow"

[lib]
path = "src/lib.rs"

[features]
default = []
diagnostics = [ "hydroflow_lang/diagnostics" ]
cli_integration = [ "hydroflow/cli_integration" ]

[dependencies]
quote = "1.0.0"
syn = { version = "2.0.0", features = [ "parsing", "extra-traits", "visit-mut" ] }
proc-macro2 = "1.0.57"
proc-macro-crate = "1.1.0"
hydroflow = { path = "../hydroflow", version = "^0.7.0", default-features = false }
hydroflow_lang = { path = "../hydroflow_lang", version = "^0.7.0" }
serde = { version = "1", features = [ "derive" ] }
bincode = "1.3"
stageleft = { path = "../stageleft", version = "^0.3.0" }
dyn-clone = "1.0.17"

# added to workaround `cargo smart-release` https://github.com/Byron/cargo-smart-release/issues/16
stageleft_tool = { path = "../stageleft_tool", version = "^0.2.0", optional = true }

[build-dependencies]
stageleft_tool = { path = "../stageleft_tool", version = "^0.2.0" }

[dev-dependencies]
insta = "1.7.1"
