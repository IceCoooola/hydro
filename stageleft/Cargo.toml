[package]
name = "stageleft"
publish = true
version = "0.3.0"
edition = "2021"
license = "Apache-2.0"
documentation = "https://docs.rs/stageleft/"
description = "Type-safe staged programming for Rust"

[lib]
path = "src/lib.rs"

[dependencies]
quote = "1.0.0"
syn = { version = "2.0.0", features = [ "parsing", "extra-traits", "visit-mut" ] }
proc-macro2 = "1.0.57"
proc-macro-crate = "1.1.0"
stageleft_macro = { path = "../stageleft_macro", version = "^0.2.0" }
