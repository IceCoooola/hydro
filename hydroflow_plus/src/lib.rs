#![feature(box_patterns)]

stageleft::stageleft_no_entry_crate!();

use std::collections::BTreeMap;
use std::marker::PhantomData;

use hydroflow::scheduled::context::Context;
pub use hydroflow::scheduled::graph::Hydroflow;
pub use hydroflow::*;
use lang::graph::{partition_graph, propagate_flow_props, HydroflowGraph};
use proc_macro2::TokenStream;
use quote::quote;
use stageleft::runtime_support::FreeVariable;
use stageleft::Quoted;

pub mod runtime_support {
    pub use bincode;
}

pub mod stream;
pub use stream::Stream;

pub mod location;
pub use location::{
    Cluster, ClusterSpec, Deploy, LocalDeploy, Location, MultiGraph, ProcessSpec,
    SingleProcessGraph,
};

pub mod cycle;
pub use cycle::HfCycle;

pub mod builder;
pub use builder::FlowBuilder;

pub mod ir;

pub mod persist_pullup;
pub mod profiler;

pub mod properties;

#[derive(Clone)]
pub struct RuntimeContext<'a> {
    _phantom: PhantomData<&'a mut &'a ()>,
}

impl Copy for RuntimeContext<'_> {}

impl<'a> FreeVariable<&'a Context> for RuntimeContext<'a> {
    fn to_tokens(self) -> (Option<TokenStream>, Option<TokenStream>) {
        (None, Some(quote!(&context)))
    }
}

pub struct HfCompiled<'a, ID> {
    hydroflow_ir: BTreeMap<usize, HydroflowGraph>,
    _phantom: PhantomData<&'a mut &'a ID>,
}

impl<'a, ID> HfCompiled<'a, ID> {
    pub fn hydroflow_ir(&self) -> &BTreeMap<usize, HydroflowGraph> {
        &self.hydroflow_ir
    }
}

impl<'a> HfCompiled<'a, usize> {
    pub fn with_dynamic_id(self, id: impl Quoted<'a, usize>) -> HfBuiltWithID<'a> {
        let hydroflow_crate = proc_macro_crate::crate_name("hydroflow_plus")
            .expect("hydroflow_plus should be present in `Cargo.toml`");
        let root = match hydroflow_crate {
            proc_macro_crate::FoundCrate::Itself => quote! { hydroflow_plus },
            proc_macro_crate::FoundCrate::Name(name) => {
                let ident = syn::Ident::new(&name, proc_macro2::Span::call_site());
                quote! { #ident }
            }
        };

        let mut conditioned_tokens = None;
        for (subgraph_id, flat_graph) in self.hydroflow_ir {
            let mut partitioned_graph =
                partition_graph(flat_graph).expect("Failed to partition (cycle detected).");

            let mut diagnostics = Vec::new();
            // Propagate flow properties throughout the graph.
            // TODO(mingwei): Should this be done at a flat graph stage instead?
            let _ = propagate_flow_props::propagate_flow_props(
                &mut partitioned_graph,
                &mut diagnostics,
            );

            let tokens = partitioned_graph.as_code(&root, true, quote::quote!(), &mut diagnostics);

            if let Some(conditioned_tokens) = conditioned_tokens.as_mut() {
                *conditioned_tokens = syn::parse_quote! {
                    #conditioned_tokens else if __given_id == #subgraph_id {
                        #tokens
                    }
                };
            } else {
                conditioned_tokens = Some(syn::parse_quote! {
                    if __given_id == #subgraph_id {
                        #tokens
                    }
                });
            }
        }

        let conditioned_tokens: TokenStream = conditioned_tokens.unwrap();
        let id = id.splice();
        HfBuiltWithID {
            tokens: syn::parse_quote!({
                let __given_id = #id;
                #conditioned_tokens else {
                    panic!("Invalid node id: {}", __given_id);
                }
            }),
            _phantom: PhantomData,
        }
    }
}

impl<'a> Quoted<'a, Hydroflow<'a>> for HfCompiled<'a, ()> {}

impl<'a> FreeVariable<Hydroflow<'a>> for HfCompiled<'a, ()> {
    fn to_tokens(mut self) -> (Option<TokenStream>, Option<TokenStream>) {
        let hydroflow_crate = proc_macro_crate::crate_name("hydroflow_plus")
            .expect("hydroflow_plus should be present in `Cargo.toml`");
        let root = match hydroflow_crate {
            proc_macro_crate::FoundCrate::Itself => quote! { hydroflow_plus },
            proc_macro_crate::FoundCrate::Name(name) => {
                let ident = syn::Ident::new(&name, proc_macro2::Span::call_site());
                quote! { #ident }
            }
        };

        if self.hydroflow_ir.len() != 1 {
            panic!("Expected exactly one subgraph in the Hydroflow IR");
        }

        let flat_graph = self.hydroflow_ir.remove(&0).unwrap();
        let mut partitioned_graph =
            partition_graph(flat_graph).expect("Failed to partition (cycle detected).");

        let mut diagnostics = Vec::new();
        // Propagate flow properties throughout the graph.
        // TODO(mingwei): Should this be done at a flat graph stage instead?
        let _ =
            propagate_flow_props::propagate_flow_props(&mut partitioned_graph, &mut diagnostics);

        let tokens = partitioned_graph.as_code(&root, true, quote::quote!(), &mut diagnostics);

        (None, Some(tokens))
    }
}

pub struct HfBuiltWithID<'a> {
    tokens: TokenStream,
    _phantom: PhantomData<&'a mut &'a ()>,
}

impl<'a> Quoted<'a, Hydroflow<'a>> for HfBuiltWithID<'a> {}

impl<'a> FreeVariable<Hydroflow<'a>> for HfBuiltWithID<'a> {
    fn to_tokens(self) -> (Option<TokenStream>, Option<TokenStream>) {
        (None, Some(self.tokens))
    }
}
