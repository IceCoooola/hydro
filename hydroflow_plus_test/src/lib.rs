stageleft::stageleft_crate!(hydroflow_plus_test_macro);

use hydroflow_plus::futures::stream::Stream;
use hydroflow_plus::tokio::sync::mpsc::UnboundedSender;
use hydroflow_plus::tokio_stream::wrappers::UnboundedReceiverStream;
use hydroflow_plus::*;
use stageleft::{q, Quoted, RuntimeData};

#[cfg(stageleft_macro)]
pub(crate) mod cluster;
#[cfg(not(stageleft_macro))]
pub mod cluster;

#[cfg(stageleft_macro)]
pub(crate) mod first_ten;
#[cfg(not(stageleft_macro))]
pub mod first_ten;

#[cfg(stageleft_macro)]
pub(crate) mod futures;
#[cfg(not(stageleft_macro))]
pub mod futures;

#[cfg(stageleft_macro)]
pub(crate) mod negation;
#[cfg(not(stageleft_macro))]
pub mod negation;

#[cfg(stageleft_macro)]
pub(crate) mod networked;
#[cfg(not(stageleft_macro))]
pub mod networked;

#[cfg(stageleft_macro)]
pub(crate) mod paxos;
#[cfg(not(stageleft_macro))]
pub mod paxos;

#[cfg(stageleft_macro)]
pub(crate) mod pbft;
#[cfg(not(stageleft_macro))]
pub mod pbft;

#[stageleft::entry(UnboundedReceiverStream<u32>)]
pub fn teed_join<'a, S: Stream<Item = u32> + Unpin + 'a>(
    flow: FlowBuilder<'a, MultiGraph>,
    input_stream: RuntimeData<S>,
    output: RuntimeData<&'a UnboundedSender<u32>>,
    send_twice: bool,
    subgraph_id: RuntimeData<usize>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let node_zero = flow.process(&());
    let node_one = flow.process(&());

    let source = flow.source_stream(&node_zero, input_stream);
    let map1 = source.clone().map(q!(|v| (v + 1, ())));
    let map2 = source.map(q!(|v| (v - 1, ())));

    let joined = map1.join(map2).map(q!(|t| t.0));

    joined.clone().for_each(q!(|v| {
        output.send(v).unwrap();
    }));

    if send_twice {
        joined.for_each(q!(|v| {
            output.send(v).unwrap();
        }));
    }

    let source_node_id_1 = flow.source_iter(&node_one, q!(0..5));
    source_node_id_1.for_each(q!(|v| {
        output.send(v).unwrap();
    }));

    flow.extract()
        .optimize_default()
        .with_dynamic_id(subgraph_id)
}

#[stageleft::entry]
pub fn chat_app<'a>(
    flow: FlowBuilder<'a, SingleProcessGraph>,
    users_stream: RuntimeData<UnboundedReceiverStream<u32>>,
    messages: RuntimeData<UnboundedReceiverStream<String>>,
    output: RuntimeData<&'a UnboundedSender<(u32, String)>>,
    replay_messages: bool,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let process = flow.process(&());

    let users = flow.source_stream(&process, users_stream).all_ticks();
    let messages = flow.source_stream(&process, messages);
    let messages = if replay_messages {
        messages.all_ticks()
    } else {
        messages.tick_batch()
    };

    // do this after the persist to test pullup
    let messages = messages.map(q!(|s| s.to_uppercase()));

    let mut joined = users.cross_product(messages);
    if replay_messages {
        joined = joined.delta();
    }

    joined.for_each(q!(|t| {
        output.send(t).unwrap();
    }));

    flow.extract().optimize_default()
}

#[stageleft::entry]
pub fn graph_reachability<'a>(
    flow: FlowBuilder<'a, SingleProcessGraph>,
    roots: RuntimeData<UnboundedReceiverStream<u32>>,
    edges: RuntimeData<UnboundedReceiverStream<(u32, u32)>>,
    reached_out: RuntimeData<&'a UnboundedSender<u32>>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let process = flow.process(&());

    let roots = flow.source_stream(&process, roots);
    let edges = flow.source_stream(&process, edges);

    let (set_reached_cycle, reached_cycle) = flow.cycle(&process);

    let reached = roots.union(reached_cycle);
    let reachable = reached
        .clone()
        .map(q!(|r| (r, ())))
        .join(edges)
        .map(q!(|(_from, (_, to))| to));
    set_reached_cycle.complete(reachable);

    reached.tick_batch().unique().for_each(q!(|v| {
        reached_out.send(v).unwrap();
    }));

    flow.extract().optimize_default()
}

#[stageleft::entry(String)]
pub fn count_elems<'a, T: 'a>(
    flow: FlowBuilder<'a, SingleProcessGraph>,
    input_stream: RuntimeData<UnboundedReceiverStream<T>>,
    output: RuntimeData<&'a UnboundedSender<u32>>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let process = flow.process(&());

    let source = flow.source_stream(&process, input_stream);
    let count = source
        .map(q!(|_| 1))
        .tick_batch()
        .fold(q!(|| 0), q!(|a, b| *a += b));

    count.for_each(q!(|v| {
        output.send(v).unwrap();
    }));

    flow.extract().optimize_default()
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use hydroflow_plus::assert_graphvis_snapshots;
    use hydroflow_plus::util::collect_ready;

    #[test]
    fn test_teed_join() {
        let (in_send, input) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut joined = super::teed_join!(input, &out, false, 0);
        assert_graphvis_snapshots!(joined);

        in_send.send(1).unwrap();
        in_send.send(2).unwrap();
        in_send.send(3).unwrap();
        in_send.send(4).unwrap();

        joined.run_tick();

        assert_eq!(&*collect_ready::<Vec<_>, _>(&mut out_recv), &[2, 3]);
    }

    #[test]
    fn test_teed_join_twice() {
        let (in_send, input) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut joined = super::teed_join!(input, &out, true, 0);
        assert_graphvis_snapshots!(joined);

        in_send.send(1).unwrap();
        in_send.send(2).unwrap();
        in_send.send(3).unwrap();
        in_send.send(4).unwrap();

        joined.run_tick();

        assert_eq!(&*collect_ready::<Vec<_>, _>(&mut out_recv), &[2, 2, 3, 3]);
    }

    #[test]
    fn test_teed_join_multi_node() {
        let (_, input) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut joined = super::teed_join!(input, &out, true, 1);
        assert_graphvis_snapshots!(joined);

        joined.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[0, 1, 2, 3, 4]
        );
    }

    #[test]
    fn test_chat_app_no_replay() {
        let (users_send, users) = hydroflow_plus::util::unbounded_channel();
        let (messages_send, messages) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut chat_server = super::chat_app!(users, messages, &out, false);
        assert_graphvis_snapshots!(chat_server);

        users_send.send(1).unwrap();
        users_send.send(2).unwrap();

        messages_send.send("hello".to_string()).unwrap();
        messages_send.send("world".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (1, "HELLO".to_string()),
                (2, "HELLO".to_string()),
                (1, "WORLD".to_string()),
                (2, "WORLD".to_string())
            ]
        );

        users_send.send(3).unwrap();

        messages_send.send("goodbye".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (1, "GOODBYE".to_string()),
                (2, "GOODBYE".to_string()),
                (3, "GOODBYE".to_string())
            ]
        );
    }

    #[test]
    fn test_chat_app_replay() {
        let (users_send, users) = hydroflow_plus::util::unbounded_channel();
        let (messages_send, messages) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut chat_server = super::chat_app!(users, messages, &out, true);
        assert_graphvis_snapshots!(chat_server);

        users_send.send(1).unwrap();
        users_send.send(2).unwrap();

        messages_send.send("hello".to_string()).unwrap();
        messages_send.send("world".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (1, "HELLO".to_string()),
                (2, "HELLO".to_string()),
                (1, "WORLD".to_string()),
                (2, "WORLD".to_string())
            ]
        );

        users_send.send(3).unwrap();

        messages_send.send("goodbye".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (3, "HELLO".to_string()),
                (3, "WORLD".to_string()),
                (1, "GOODBYE".to_string()),
                (2, "GOODBYE".to_string()),
                (3, "GOODBYE".to_string())
            ]
        );
    }

    #[test]
    pub fn test_reachability() {
        let (roots_send, roots) = hydroflow_plus::util::unbounded_channel();
        let (edges_send, edges) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut reachability = super::graph_reachability!(roots, edges, &out);
        assert_graphvis_snapshots!(reachability);

        roots_send.send(1).unwrap();
        roots_send.send(2).unwrap();

        edges_send.send((1, 2)).unwrap();
        edges_send.send((2, 3)).unwrap();
        edges_send.send((3, 4)).unwrap();
        edges_send.send((4, 5)).unwrap();

        reachability.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[1, 2, 3, 4, 5]
        );
    }

    #[test]
    pub fn test_count() {
        let (in_send, input) = hydroflow_plus::util::unbounded_channel();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut count = super::count_elems!(input, &out);
        assert_graphvis_snapshots!(count);

        in_send.send(1).unwrap();
        in_send.send(1).unwrap();
        in_send.send(1).unwrap();

        count.run_tick();

        assert_eq!(&*collect_ready::<Vec<_>, _>(&mut out_recv), &[3]);
    }
}
