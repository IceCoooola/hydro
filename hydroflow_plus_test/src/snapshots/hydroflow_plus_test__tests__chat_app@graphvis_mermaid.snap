---
source: hydroflow_plus_test/src/lib.rs
expression: "chat_server.meta_graph().unwrap().to_mermaid(&Default::default())"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1[\"<div style=text-align:center>(1v1)</div> <code>source_stream({<br>    use crate::__staged::*;<br>    let users_stream = users_stream;<br>    users_stream<br>})</code>"/]:::pullClass
2v1[\"(2v1) <code>tee()</code>"/]:::pullClass
3v1[\"(3v1) <code>persist()</code>"/]:::pullClass
4v1[\"(4v1) <code>tee()</code>"/]:::pullClass
5v1[\"<div style=text-align:center>(5v1)</div> <code>source_stream({<br>    use crate::__staged::*;<br>    let messages = messages;<br>    messages<br>})</code>"/]:::pullClass
6v1[\"(6v1) <code>tee()</code>"/]:::pullClass
7v1[\"(7v1) <code>cross_join::&lt;'tick, 'tick&gt;()</code>"/]:::pullClass
8v1[\"(8v1) <code>tee()</code>"/]:::pullClass
9v1[/"<div style=text-align:center>(9v1)</div> <code>for_each({<br>    use crate::__staged::*;<br>    let output = output;<br>    |t| {<br>        output.send(t).unwrap();<br>    }<br>})</code>"\]:::pushClass
1v1-->2v1
3v1-->4v1
2v1-->3v1
5v1-->6v1
7v1-->8v1
4v1-->|0|7v1
6v1-->|1|7v1
8v1-->9v1
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1
    2v1
    3v1
    4v1
    5v1
    6v1
    7v1
    8v1
    9v1
    subgraph sg_1v1_var_stream_0 ["var <tt>stream_0</tt>"]
        1v1
        2v1
    end
    subgraph sg_1v1_var_stream_1 ["var <tt>stream_1</tt>"]
        3v1
        4v1
    end
    subgraph sg_1v1_var_stream_2 ["var <tt>stream_2</tt>"]
        5v1
        6v1
    end
    subgraph sg_1v1_var_stream_3 ["var <tt>stream_3</tt>"]
        7v1
        8v1
    end
    subgraph sg_1v1_var_stream_4 ["var <tt>stream_4</tt>"]
        9v1
    end
end

